### PR TITLE
cmd: add non-stub implementation of snap-update-ns

### PR DESCRIPTION
**Pull request by zyga**
This patch adds an almost full implementation of snap-update-ns.

This new executable includes a simple utility module for working with
mount entries (e.g. /etc/fstab) that is based on the C library for the
hard part (parsing).

The missing bits are:
 - there's no locking; I will tweak ns-support to expose that better
 - snapd doesn't generate the snap.$SNAP_NAME.fstab file yet
 - snapd doesn't call snap-update-ns yet (easy)
 - snap-update-ns cannot yet realize that a mount already happened

The last missing bit requires some more explanation. The tool already
inspect existing mount points but due to the way bind mounts are exposed
in the kernel it is non-trivial to realize that a given place is a bind
mount (even if we have all the information in front of us).

One notable unfortunate fact is that snap-update-ns cannot take advantage
of valgrind for debugging:

--41370-- WARNING: unhandled amd64-linux syscall: 308
--41370-- You may be able to write your own handler.
--41370-- Read the file README_MISSING_SYSCALL_OR_IOCTL.
--41370-- Nevertheless we consider this a bug.  Please report
--41370-- it at http://valgrind.org/support/bug_reports.html.
cannot re-associate with mount namespace of snap unix-domain-client: Function not implemented

Signed-off-by: Zygmunt Krynicki <zygmunt.krynicki@canonical.com>

### CODE REVIEW

File: cmd/snap-update-ns/snap-update-ns.c [hunk 1]
```diff
@@ -19,7 +19,299 @@
...
+			debug("\tActing on the change...");
+			sc_act_on_mount_change(&change);
+			num_changed += 1;
+		} else {
+			debug("\tNot acting on the change!");
+			num_skipped += 1;
+		}
+	}
+	if (num_skipped > 0) {
+		debug("Mount namespace of snap %s has not been fully altered.",
```
Comment (jdstrand) [inline cmd/snap-update-ns/snap-update-ns.c]: s/altered/updated/ here and elsewhere

File: cmd/snap-update-ns/snap-update-ns.c [hunk 2]
```diff
@@ -19,7 +19,299 @@
...
+	if (current == NULL) {
+		debug("\tcurrent: (empty profile)");
+	}
+
+	if (current == NULL && desired == NULL) {
+		debug("There's nothing to do");
+		return 0;
+	}
+	// TODO: correct the wiki, we don't quit if something is not present as
+	// this is a valid case as well (e.g. a profile gets removed).
```
Comment (jdstrand) [inline cmd/snap-update-ns/snap-update-ns.c]: The use cases are:
- neither exists: nothing to do
- current exists but desired does not: unmount everything and remove current (profile removal)
- current does not exist, but desired does: mount everything and create current (first run after profile add)
- both exist: mount/umount as necessary and update current (subsequent updates not including profile removal)

File: cmd/snap-update-ns/snap-update-ns.c (hunk 1 above)
Comment (zyga) [inline cmd/snap-update-ns/snap-update-ns.c]: Yep, will do.

File: cmd/snap-update-ns/snap-update-ns.c [hunk 3]
```diff
@@ -19,7 +19,299 @@
...
+
+static bool sc_should_act_on_change(const struct sc_mount_change *change)
+{
+	// Load the table of mount points that affect the current process.  We're
+	// doing this each time we are asked to mount something as it is safer than
+	// trying to keep track of what the kernel may be doing.
+	struct mountinfo __attribute__ ((cleanup(cleanup_mountinfo))) * mi =
+	    NULL;
+	const char *mnt_dir;
+	mi = parse_mountinfo(NULL);
```
Comment (zyga) [inline cmd/snap-update-ns/snap-update-ns.c]: Good catch, will do

File: cmd/snap-update-ns/snap-update-ns.c [hunk 4]
```diff
@@ -19,7 +19,299 @@
...
+							    (mi_entry));
+				while (parent_mi_entry != NULL) {
+					debug("\t(parent chain)...");
+					sc_show_mountinfo(parent_mi_entry);
+					parent_mi_entry =
+					    sc_find_mountinfo_by_id(mi,
+								    mountinfo_entry_parent_id
+								    (parent_mi_entry));
+				}
+				return false;
```
Comment (zyga) [inline cmd/snap-update-ns/snap-update-ns.c]: I'll trim down the debug info if required (though we don't show it unless you ask for it). I'd like to return a simple and understandable error message when we reject the over-mount.

File: cmd/snap-update-ns/snap-update-ns.c [hunk 5]
```diff
@@ -19,7 +19,299 @@
...
+static bool sc_should_act_on_change(const struct sc_mount_change *change);
+
+#define SC_DESIRED_PROFILE_FMT "/var/lib/snapd/mount/snap.%s.fstab"
+#define SC_CURRENT_PROFILE_FMT "/run/snapd/ns/%s.fstab"
+#define SC_MNT_NS_FMT          "/run/snapd/ns/%s.mnt"
+
+int main(int argc, char **argv)
+{
+	if (argc != 2) {
+		printf("Usage: snap-alter-ns SNAP-NAME");
```
Comment (zyga) [inline cmd/snap-update-ns/snap-update-ns.c]: Yep, will do.

File: cmd/snap-update-ns/snap-update-ns.c (hunk 2 above)
Comment (zyga) [inline cmd/snap-update-ns/snap-update-ns.c]: - neither exist == both are empty => nothing to do (check, agreed)
- current exists but desired does not == treat desired as empty => will unmount everything in current (check, agreed)
- current does not exist but desired does == treat current as empty => will mount everything in desired (check, agreed with the exception that I plan to make snap-confine write current on successful startup)
- both exists: unmount those that are not in desired but are in current, mount those that are in desired but are not in current; (check, agreed)

I think we are in agreement here.

**Comment by zyga** [comment]
I'll address the review feedbackc and focus on investigating the kernel issue.

**Comment by niemeyer** [comment]
For the record, @zyga is using this as a master sketch for proposing other PRs. Closing it as it'll eventually be all done in smaller pieces, which is much better. Thanks for that @zyga.

**Pull request closed without merging**